embedding/transformation/lightning_transformation/question_answering_output_transformation.py

```diff
@@ -0,0 +1,218 @@
+
+    for batch_id, batch in enumerate(predictions):
+        for key in batch.keys():
+            if key not in tensors_lists_dict.keys():
+                tensors_lists_dict[key] = defaultdict(list)
+            for tensor_key, tensor in batch[key].items():
+                if tensor_key == "loss":
+                    continue
+                assert isinstance(tensor, torch.Tensor)
+                if tensor.dtype in {torch.bfloat16, torch.float16}:
+                    tensor = tensor.to(dtype=torch.float32)
+                tensors_lists_dict[key][tensor_key].append(tensor)
+
+    output: Dict[str, Dict[str, torch.Tensor]] = defaultdict(dict)
+
+    for key in tensors_lists_dict.keys():
+        for tensor_key, tensors in tensors_lists_dict[key].items():
+            output[key][tensor_key] = torch.cat(tensors)
+
+    return output
+
+
+class QABasePostprocessor(abc.ABC):
+    @abc.abstractmethod
+    def postprocess(
+        self,
+        examples: Dataset,
+        overflow_to_sample_mapping: List[int],
+        offset_mappings: List[List[List[int]]],
+        outputs: List[Dict[str, Any]],
+    ) -> List[Dict[str, Any]]:
+        pass
+
+
+class QAPredictionPostProcessor(QABasePostprocessor):
+    """
+    Based on QA huggingface transformers pipeline.
+    https://github.com/huggingface/transformers/blob/d6b8e9cec7301ba02f642588a6f12e78ec3b9798/examples/pytorch/question-answering/utils_qa.py#L31
+    https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/question_answering.ipynb#scrollTo=n9qywopnIrJH
+    """
+
+    def __init__(
+        self,
+        top_k_answers: int = 20,
+        max_answer_length: int = 40,
+        question_field_name: str = "question",
+        context_field_name: str = "context",
+        target_field_name: str = "answers",
+    ) -> None:
+        super().__init__()
+        self.top_k_answers = top_k_answers
+        self.max_answer_length = max_answer_length
+        self.question_field_name = question_field_name
+        self.context_field_name = context_field_name
+        self.target_field_name = target_field_name
+
+    def _get_topk_not_cls_predictions_from_output(
+        self, start_logits: torch.Tensor, end_logits: torch.Tensor, offset_mapping: List[List[int]]
+    ) -> List[Dict[str, Any]]:
+
+        topk_start_indices = torch.topk(start_logits, self.top_k_answers).indices.tolist()
+        topk_end_indices = torch.topk(end_logits, self.top_k_answers).indices.tolist()
+        topk_predictions = []
+
+        for start_index in topk_start_indices:
+            start_index_offset = offset_mapping[start_index + 1]
+
+            for end_index in topk_end_indices:
+                end_index_offset = offset_mapping[end_index + 1]
+
+                # Don't consider out-of-scope answers, either because the indices are out of bounds or correspond
+                # to part of the input_ids that are not in the context.
+                if (
```

Comment:
Was it a problem with the dataset? it looks more like preprocessing/filtering code